javascript/newer-JS-features.md

```diff
@@ -0,0 +1,219 @@
+
+## Spread Operator
+The Spread (...) operator is a useful and newer syntax for adding items to arrays, combining arrays or objects, and spreading an array out into a function’s arguments.
+## Some scenarios where Spread is useful:
+- Copying an array
+- Conncatenating or combining arrays
+- Merge two Objects.
+
+## Examples
+### Spread in Function Calls
+We can use Math.max to get the greatest number from a list.
+```javascript
+Math.max(2,3,4,5,6) // 6
+```
+Here, Math.max ia expecting seperate arguments or values to be passed on and it'll give us the maximum value. But, what if we use Math.max on array?
+```javascript
+Math.max([2,3,4,5]); // Nan
+```
+Here, we got NaN because Math.max is expecting seperate values each one being a number. Luckily we have Spread to the rescue. Spread syntax uses three dots ... to spread the array as seperate arguments.
+
+```javascript
+Math.max(...[2,3,4]); // 4 (spread expands array into seperate arguments)
+```
+### Spread with Arrays
+Spread in array literals creates a new array using an existing array. It basically spreads the elements from one array into a new array.
+Let's look an example -
+Here, I've 2 arrays primaryColors and secondaryColors. allColors array contain all the elemnts from primaryColors and secondaryColors. ... spread the values into seperate arguments. allColors contains copy of primaryColors and secondarycolors but the original arrays remains unchanged! 
+```javascript
+const primaryColors = ['red', 'green', 'blue'];
+const secondaryColors = ['purple','pink', 'crimson'];
+const allColors = [...primaryColors, ...secondaryColors] // (6)["red", "green", "blue", "purple", "pink", "crimson"]
+```
+### Spread with Objects
+Spread in object literal copies properties from one object into another object. In past, we've used Object.assign to copy properties from one object to another. Spread makes copying more easy!
+```javascript
+const personA = {
+    favColor:"pink"
+}
+const personB = {
+    favFood:"pasta"
+}
+const info = {...personA, ...personB}
+console.log(info); // {favColor: "pink", favFood: "pasta"}
+```
+In case of conflict where both objects have similar properties, order matters! 
+```javascript
+const personC = {
+    favColor:"pink",
+    favPet:"cat"
+}
+const personD = {
+    favFood:"pasta",
+    favPet:"dog"
+}
+const anotherInfo = {...personA, ...personB } // {favColor: "pink", favPet: "dog", favFood: "pasta"}
+```
+## Rest parameters...
+Rest parameters allows a function to accept an indefinite numbers of arguments as an array. Rest looks like spread, but it’s NOT spread. Before taking a look at Rest, let's take a look at **Arguments Object** which  is an array like object but it doesn't have access to array like methods like map, reduce. Also, it's not available with arrow function. It automatically hold all the values passed to the function.
```

Comment:
Is "rest" capitalized?  I think it may make more sense to mark it as code. Basically, I would be careful how you annotate things, because this is likely the first time the learner has heard about it, and they will be trying to internalize these things. Is is a Proper Name, a `code method` to remember, a _term_? 
```suggestion
`rest` parameters allows a function to accept an indefinite numbers of arguments as an array. `rest` looks like spread, but it’s NOT spread. Before taking a look at `rest`, let's take a look at [`arguments` object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/arguments)  which  is an array like an object but it doesn't have access to array like methods like `map()` and `reduce()`. Also, it's not available with an _arrow function_. It automatically holds all the values passed to the function.
```